ciongke/cmd/test-pr/main.go

```diff
@@ -0,0 +1,392 @@
+	KubeClient      client.Interface
+	KubeBatchClient client.BatchInterface
+}
+
+func main() {
+	flag.Parse()
+
+	oauthSecret, err := ioutil.ReadFile(*oauthSecretFile)
+	if err != nil {
+		log.Fatalf("Could not read oauth secret file: %s", err)
+	} else if len(oauthSecret) == 0 {
+		log.Fatalf("OAuth secret must not be empty.")
+	}
+
+	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: string(oauthSecret)})
+	tc := oauth2.NewClient(oauth2.NoContext, ts)
+	githubClient := github.NewClient(tc)
+
+	clonePath := filepath.Join(*workspace, "repo")
+	mergeable, err := checkoutPR(*repoURL, clonePath, *branch, *head, *pr)
+	if err != nil {
+		log.Printf("Error preparing git repo: %s", err)
+		return
+	}
+	if !mergeable {
+		log.Printf("PR %d needs rebase, skipping tests.", *pr)
+		return
+	}
+
+	sourcePath := filepath.Join(*workspace, *head+".tar.gz")
+	tar := exec.Command("tar", "czf", sourcePath, "repo")
+	tar.Dir = *workspace
+	if err = cmdlog.RunWithLogs(tar); err != nil {
+		log.Printf("Failed to create source archive: %s", err)
+		return
+	}
+	if err = gcs.UploadSource(*sourceBucket, *head, sourcePath); err != nil {
+		log.Printf("Failed to upload source: %s", err)
+		return
+	}
+
+	b, err := ioutil.ReadFile(filepath.Join(clonePath, ".test.yml"))
+	if err != nil {
+		log.Printf("Cannot read .test.yml: %s", err)
+		return
+	}
+	jobs := []jobDescription{}
+	if err = yaml.Unmarshal(b, &jobs); err != nil {
+		log.Printf("Cannot unmarshal .test.yml: %s", err)
+		return
+	}
+
+	config, err := restclient.InClusterConfig()
+	if err != nil {
+		log.Fatalf("Error getting rest client: %s", err)
+	}
+	cl, err := client.New(config)
+	if err != nil {
+		log.Fatalf("Error getting client: %s", err)
+	}
+	batchClient, err := client.NewBatch(config)
+	if err != nil {
+		log.Fatalf("Error getting batch client: %s", err)
+	}
+	testClients := testClient{
+		GitHubClient:    githubClient,
+		KubeClient:      cl,
+		KubeBatchClient: batchClient,
+	}
+
+	var wg sync.WaitGroup
+	for _, job := range jobs {
+		wg.Add(1)
+		go test(testClients, job, &wg)
+	}
+	wg.Wait()
+}
+
+// checkoutPR does the checkout and returns whether or not the PR can be merged.
+func checkoutPR(repoPath, clonePath, branch, head string, prNumber int) (bool, error) {
+	cloneCommand := exec.Command("git", "clone", repoPath, clonePath)
+	checkoutCommand := exec.Command("git", "checkout", branch)
+	fetchCommand := exec.Command("git", "fetch", "origin", "pull/"+strconv.Itoa(prNumber)+"/head:pr")
+	mergeCommand := exec.Command("git", "merge", head, "--no-edit")
+	if err := cmdlog.RunWithLogs(cloneCommand); err != nil {
+		return false, err
+	}
+	checkoutCommand.Dir = clonePath
+	if err := cmdlog.RunWithLogs(checkoutCommand); err != nil {
+		return false, err
+	}
+	fetchCommand.Dir = clonePath
+	if err := cmdlog.RunWithLogs(fetchCommand); err != nil {
+		return false, err
+	}
+	mergeCommand.Dir = clonePath
+	if err := cmdlog.RunWithLogs(mergeCommand); err != nil {
+		return false, nil
```

Comment:
If the merge command fails then it couldn't merge and I want to treat that differently than if there's a problem cloning, for instance.